Source/Session.swift

```diff
@@ -469,21 +482,33 @@ open class Session {
     }
 
     func updateStatesForTask(_ task: URLSessionTask, request: Request) {
-        switch (startRequestsImmediately, request.state) {
-        case (true, .initialized):
-            request.resume()
-        case (false, .initialized):
-            // Do nothing.
-            break
-        case (_, .resumed):
-            task.resume()
-            request.didResumeTask(task)
-        case (_, .suspended):
-            task.suspend()
-            request.didSuspendTask(task)
-        case (_, .cancelled):
-            task.cancel()
-            request.didCancelTask(task)
+        requestStateQueue.sync {
+            switch (startRequestsImmediately, request.state) {
+            case (true, .initialized):
+                guard request.attemptToTransitionTo(.resumed) else { return }
+
+                request.didResume()
+
+                task.resume()
+                request.didResumeTask(task)
```

Comment:
Should we hop back to `rootQueue` for these `didResume` and `didResumeTask` calls as well? The only things we actually need to call on `requestStateQueue` are the transition and `task.resume()`.